traits/trait_list_object.py

```diff
@@ -53,465 +90,765 @@ def __init__(self, index=0, removed=None, added=None):
+
+        super().__init__(value, validator=self.trait_validator,
+                         notifiers=[self.notifier] + notifiers)
+
+    def trait_validator(self, trait_list, index, removed, added):
+        """ Validates the value by calling the inner trait's validate method
+        and also ensures that the size of the list is within the specified
+        bounds.
+
+        Parameters
+        ----------
+        trait_list : list
+            The current list.
+        index : index or slice
+            Index or slice corresponding to the values added/removed.
+        removed : list
+            values that are removed.
+        added : list
+            value or list of values that are added.
+
+        Returns
+        -------
+        value : validated value or list of validated values.
+
+        Raises
+        ------
+        TraitError
+            On validation failure for the inner trait or if the size of the
+            list exceeds the specified bounds
+        """
+        object_ref = getattr(self, 'object', None)
+        trait = getattr(self, 'trait', None)
+
+        if object_ref is None or trait is None:
+            return added
+
+        object = object_ref()
+
+        # check that length is within bounds
+        new_len = len(trait_list) - len(removed) + len(added)
+        if not trait.minlen <= new_len <= trait.maxlen:
+            raise TraitError(
+                "The '%s' trait of %s instance must be %s, "
+                "but you attempted to change its length to %d element%s."
+                % (
+                    self.name,
+                    class_of(object),
+                    self.trait.full_info(object, self.name, Undefined),
+                    new_len,
+                    "s"[new_len == 1:],
+                )
             )
+
+        # validate the new value(s)
+        validate = trait.item_trait.handler.validate
+        if validate is None or added == []:
+            return added
+
+        validate = adapt_trait_validator(validate, self.name)
+        return validate(object, index, removed, added)
+
+    def notifier(self, trait_list, index, removed, added):
+        """ Converts and consolidates the parameters to a TraitListEvent and
+        then fires the event.
+
+        Parameters
+        ----------
+        trait_list : list
+            The list
+        index : int or slice
+            Index or slice that was modified
+        removed : list
+            Values that were removed
+        added : list
+            Values that were added
+
+        """
+        is_trait_none = self.trait is None
+        is_name_items_none = self.name_items is None
+        if not hasattr(self, "trait") or is_trait_none or is_name_items_none:
+            return
+
+        object = self.object()
+        if object is None:
+            return
+
+        # bug-for-bug conversion of parameters to TraitListEvent
+        if isinstance(index, slice):
+            if index.step in {1, None}:
```

Comment:
nit: spell this as `if index.step is None or index.step == 1` (or something like that); it's normal to compare with `None` using `is` rather than equality